### PR TITLE
Add ObjectBox Vector Store Integration

### PR DESCRIPTION
**Pull request by shubham0204**
# Description

This PR adds a vectorstore integration for ObjectBox.
[ObjectBox](https://objectbox.io/) is a fast, efficient on-device NoSQL database that supports vector-indexing. 
New dependencies are [objectbox](https://pypi.org/project/objectbox/0.6.0/).

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/poetry.lock
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/poetry.lock:1]: we can remove this file

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,140 @@
+<<<<<<< Updated upstream
```
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml:1]: remnants of an old merge conflict?

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+version = "<=23.9.1,>=23.7.0"
+
+[tool.poetry.group.dev.dependencies.codespell]
+extras = ["toml"]
+version = ">=v2.2.6"
+
+[[tool.poetry.packages]]
+include = "llama_index/"
+||||||| Stash base
+=======
```
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml:71]: More remnants

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/README.md
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/README.md:1]: Nice readme! ❤️

**Review by logan-markewich — APPROVED** [review]
Looks great honestly! Just a few nits before we can merge it

**Comment by shubham0204** [comment]
@logan-markewich I have resolved the issues you pointed out👍🏻

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml (hunk 1 above)
Comment (shubham0204) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml]: Yes, it has been removed now

File: llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml (hunk 2 above)
Comment (shubham0204) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/pyproject.toml]: I should have checked this file 😓 . But anyways, I have removed them 👍🏻

**Comment by logan-markewich** [comment]
@shubham0204 great, looks good.

Sadly, I can't push to your PR / branch, but there's a few things to clean up
- try running `make lint` to automatically apply code formatting (or if it can't format, it will report errors)
- you'll need to add/edit the following file: `lama-index-integrations/vector_stores/llama-index-vector-stores-objectbox/BUILD`

It will have exactly these contents
```
poetry_requirements(
    name="poetry",
)

```

**Comment by shubham0204** [comment]
@logan-markewich I have added the `BUILD` file with the required contents. Also, I ran `make lint`, and fixed the errors it gave 👍🏻

**Comment by logan-markewich** [comment]
@shubham0204 seems like the tests are failing -- did you try running them locally?

Also, it seems like linting is still failing -- try running `make lint` from the root of the repo, it will only lint files/folders underneath from where it is invoked

**Comment by shubham0204** [comment]
@logan-markewich there was an issue with a filepath in the tests (they were running locally when executed from the root of the ObjectBox integration directory) which has been fixed now. Also, the `ObjectBoxIndexDemo.ipynb` notebook is not formatted according to the linter 👍🏻

**Comment by logan-markewich** [comment]
@shubham0204 Did you publish this package on your own? It failed to publish in our cicd

`HTTP Error 403: The user 'jerryjliu' isn't allowed to upload to project 'llama-index-vector-stores-objectbox'`

I really do not recommend publishing on your own -- I think you'd have to either transfer ownership, or handle publishing updates yourself

**Comment by shubham0204** [comment]
> @shubham0204 Did you publish this package on your own? It failed to publish in our cicd
> 
> `HTTP Error 403: The user 'jerryjliu' isn't allowed to upload to project 'llama-index-vector-stores-objectbox'`
> 
> I really do not recommend publishing on your own -- I think you'd have to either transfer ownership, or handle publishing updates yourself

@logan-markewich Yes, the package has been published by ObjectBox. Transferring the ownership of the package may not be possible. Is there any workaround or "rights" that can be given to LlamaIndex to publish the package?

**Comment by logan-markewich** [comment]
@shubham0204 no idea actually -- not a pypi expert. This has only happened once before and we were able to transfer 

To be clear, its totally fine, but ObjectBox will be responsible for publishing then

**Pull request merged**